asteroid/data/sms_wsj_dataset.py

```diff
@@ -0,0 +1,219 @@
+        dataset = db.get_dataset(dset)
+        # Filter out short utterances only when segment is specified
+        orig_len = len(dataset)
+        drop_utt, drop_len = 0, 0
+        if not self.like_test:
+            def filter_short_examples(example):
+                num_samples = example['num_samples']['observation']
+                if num_samples < self.seg_len:
+                    return False
+                else:
+                    return True
+
+            dataset = dataset.filter(filter_short_examples, lazy=False)
+        self.dataset = dataset
+
+    def __add__(self, sms_wsj):
+        if self.n_src != sms_wsj.n_src:
+            raise ValueError('Only datasets having the same number of sources'
+                             'can be added together. Received '
+                             '{} and {}'.format(self.n_src, sms_wsj.n_src))
+        if self.seg_len != sms_wsj.seg_len:
+            self.seg_len = min(self.seg_len, sms_wsj.seg_len)
+            print('Segment length mismatched between the two Dataset'
+                  'passed one the smallest to the sum.')
+        self.dataset = self.dataset.concatenate(sms_wsj.dataset)
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __getitem__(self, idx):
+        """ Gets a mixture/sources pair.
+        Returns:
+            mixture, vstack([source_arrays])
+        """
+        # Random start
+        example = self.dataset[idx]
+        in_signal = self.target_dict['mixture']
+        target = self.target_dict['target']
+        audio_path = example['audio_path']
+        num_samples = example['num_samples']['observation']
+        if num_samples == self.seg_len or self.like_test:
+            rand_start = 0
+        else:
+            rand_start = np.random.randint(0, num_samples - self.seg_len)
+        if self.like_test:
+            stop = None
+        else:
+            stop = rand_start + self.seg_len
+        # Load mixture
+        x, _ = sf.read(audio_path[in_signal], start=rand_start,
+                       stop=stop, dtype='float32')
+        x = x.T
+
+        num_channels = self.target_dict['infos']['num_channels']
+        if self.single_channel:
+            if self.like_test:
+                ref_channel = 0
+            else:
+                ref_channel = np.random.randint(0, num_channels)
+            x = x[ref_channel]
+        seg_len = torch.as_tensor([x.shape[-1]])
+        # Load sources
+        source_arrays = []
+        for idx in range(self.n_src):
+            try:
+                s = 0
+                for t in target:
+                    if t == 'speech_source':
+                        start = 0
+                        stop_ = None
+                    else:
+                        start = rand_start
+                        stop_ = stop
+                    s_, _ = sf.read(audio_path[t][idx], start=start,
+                                    stop=stop_, dtype='float32')
+                    s += s_.T
```

Comment:
Clarification, this sum is to handle this case : ` speech_image = ['speech_reverberation_early', 'speech_reverberation_tail']` where the training target is the whole reverberated signal, is that right?

Reply:
yes that is right. To minimize the disk space required we only write down the signals which cannot easily be generated from other signals. Therefore, the whole reverberated signal has to recalculated. I will add a comment